### ISSUE TITLE
Comments containing > character.

### ISSUE THREAD

**Issue by emkw** [Ygg01/xml5_draft#8]
Regarding interoperability with XML 1.0 and 1.1 for XML5 - comments containing `>` chars in the middle are valid in XML 1.0 and 1.1. So the change introduced in 2caadd3ea0296e9e0706a03d2a0cfe36d80bb7ac may cause break on those.

This example is in both XML 1.0 and 1.1 spec, as a valid comment:

``` xml
<!-- declarations for <head> & <body> -->
```

Was error on `>` character in comment introduced to overcome some class of bugs?

**Comment by Ygg01** [comment]
Hm a valid point. I based XML 5.0 to be compatible with HTML spec. However this seems like a case where it might be better to be more compliant with XML spec. 

Anyway, I will be off on vacation, so I'll see updating this, tests and parser.

@annevk got any ideas/comments about this issue?

**Comment by annevk** [comment]
I think we should not be incompatible with XML. We can only "innovate" when the XML would not be well-formed anyway.

**Comment by Ygg01** [comment]
Well I looked into it, and while it's easy to remove that rule, it changes how we parse errors in comments. E.g.

`<!-->` becomes  `ParseErr, Comment(">")`
`<!-->test` becomes `ParseErr, Comment(">test")`

while

`<!>` remains `ParseErr, Comment("")` 

One solution to this "problem" is to recognize that after `<!--` any `>` ends comment, but then comments like `<!-->>Text -->` aren't valid.

**Comment by sideshowbarker** [comment]
To get back to [the OP’s question](https://github.com/Ygg01/xml5_draft/issues/8#issue-167185344):

> This example is in both XML 1.0 and 1.1 spec, as a valid comment:
> 
>   `<!-- declarations for <head> & <body> -->`
> 
> Was error on `>` character in comment introduced to overcome some class of bugs?

About [the following response](https://github.com/Ygg01/xml5_draft/issues/8#issuecomment-234773603): 

> Hm a valid point. I based XML 5.0 to be compatible with HTML spec. However this seems like a case where it might be better to be more compliant with XML spec.

The comment `<!-- declarations for <head> & <body> -->` is actually also a conforming comment per the HTML spec. As far as I understand, it has always been a conforming HTML comment. So this isn’t a case where the requirements in XML are different from HTML, right?

**Comment by Ygg01** [comment]
> As far as I understand, it has always been a conforming HTML comment. So this isn’t a case where the requirements in XML are different from HTML, right?

Unsure. It seems like the spec changed somewhat in meantime (I remember `CommentDash` being in spec and now its not?!) or my mind is playing tricks at me. I'll add new comment states and see how it goes. This would be ideal solution.

**Comment by sideshowbarker** [comment]
> It seems like the spec changed somewhat in meantime (I remember `CommentDash` being in spec and now its not?!) or my mind is playing tricks at me.

Yeah, two changes made recently:
- [Allow dashes in comments](https://github.com/whatwg/html/commit/518d16fdc672d1023dcfd2847d86f559d13a842f)
- [Allow comments to end with `<!-->`](https://github.com/whatwg/html/commit/1633b9aadab815e6daba17950c4ff81a80b896ed)

But neither of those relates to the `<!-- declarations for <head> & <body> -->` case (and incidentally neither changes how those cases are actually parsed; instead both only affect the optional [“parse error”](https://html.spec.whatwg.org/multipage/syntax.html#parse-error)-reporting behavior).

**Comment by Ygg01** [comment]
Yeah. This is a big error either way. It's both non XML compliant and different from the way HTML5 does things. 

The patch is on my machine, and I'll update it later this evening.

**Comment by ExE-Boss** [comment]
Note that **XML 1.0** allows: `<!-- comment with > and ending with <!-->`.

It’s only `--` that’s not allowed in **XML 1.0** comments
